lib/ws-kernel-picker.coffee

```diff
@@ -0,0 +1,118 @@
+{SelectListView} = require 'atom-space-pen-views'
+_ = require 'lodash'
+
+Config = require './config'
+services = require './jupyter-js-services-shim'
+WSKernel = require './ws-kernel'
+uuid = require 'uuid'
+
+class CustomListView extends SelectListView
+    initialize: (@emptyMessage, @onConfirmed) ->
+        super
+        @addClass('watch-language-picker')
+        @storeFocusedElement()
+        @panel ?= atom.workspace.addModalPanel(item: this)
+        @panel.show()
+        @focusFilterEditor()
+
+    getFilterKey: ->
+        'name'
+
+    destroy: ->
+        @cancel()
+
+    viewForItem: (item) ->
+        element = document.createElement('li')
+        element.textContent = item.name
+        element
+
+    cancelled: ->
+        @panel?.destroy()
+        @panel = null
+        @editor = null
+
+    confirmed: (item) ->
+        @onConfirmed?(item)
+        @cancel()
+
+    getEmptyMessage: ->
+        @emptyMessage
+
+module.exports =
+class WSKernelPicker
+    constructor: (onChosen) ->
+        @_onChosen = onChosen
+
+    toggle: ->
+        @_path = atom.workspace.getActiveTextEditor().getPath() + '-' + uuid.v4()
+        gatewayListing = new CustomListView('No gateways available', @onGateway.bind this)
+        gatewayListing.setLoading 'Loading gateways...'
+        items = Config.getJson('gateways', [])
+        if not _.isEmpty items
+            gatewayListing.setItems items
+            gatewayListing.setError 'Select a gateway' # TODO(nikita): maybe don't misuse error
+        else
+            gatewayListing.setError 'Use the package settings to specify gateway information'
+
+    onGateway: (spec) ->
+        console.log('Picked a gateway')
+        sessionListing = new CustomListView('No sessions available', @onSession.bind this)
+        sessionListing.setLoading 'Loading sessions...'
+
+        services.listRunningSessions(spec.options)
+        .then (sessionModels) ->
+            items = sessionModels.map (model) ->
+                if model.notebook?.path?
+                    name = model.notebook.path
+                else
+                    name = "Session #{model.id}"
+                return {
+                    name: name
+                    model: model
+                    options: spec.options
+                }
+            items.unshift
+                name: '[new session]'
+                model: null
+                options: spec.options
+            sessionListing.setItems(items)
+
+        , (err) =>
+            # Gateways offer the option of never listing sessions, for security
+            # reasons.
+            # Assume this is the case and proceed to creating a new session.
```

Comment:
I'm confused about your question here, because the possibility of creating a new session exists in this very file!